### ISSUE TITLE
[BITV] 9.1.3.1e/8.1 - The user table represents a grid element and should be implemented as such. (1)

### ISSUE THREAD

**Issue by AndyScherzinger** [nextcloud/server#36921]
### ⚠️⚠️⚠️ Regression after table implementation: https://github.com/nextcloud/server/pull/38424

At the moment, the data is implemented in an unstructured manner using "div" elements. More information on grid implementation can be found at the following URLs: https://sarahmhigley.com/writing/grids-part1/, https://sarahmhigley.com/writing/grids-part2/ or https://www.w3.org/WAI/ARIA/apg/patterns/grid/. 

- [x] https://github.com/nextcloud/server/pull/39050
- ~https://github.com/Akryum/vue-virtual-scroller/pull/825~

### If upstream additions can't be made
- [x] https://github.com/nextcloud/server/pull/40719

![4b98244143224a248f230b7ff75c24c8](https://user-images.githubusercontent.com/1315170/221935104-35e9ce6c-e80e-44c8-b685-68b191af0e77.png)

<details>
<summary>Details</summary>

https://report.bitvtest.de/default-en/d63601ac-cb34-4645-8256-66bec78964a0.html#checkpoint-3f0ff0bc0e-v8-n1

</details>

**Comment by JuliaKirschenheuter** [comment]
Hi @michaelnissenbaum 
I would like to ask you about your opinion that this overview should be a grid and not a table. 

> Possible user actions within a grid could include editing a cell value (common in a spreadsheet), selecting or deselecting a cell, and reordering, inserting, and deleting rows or columns. If most of the cells are not interactive, then it seems likely that a table would be a better choice than a grid.
https://sarahmhigley.com/writing/grids-part1/

In my opinion the most of the cell aren't interactive: only 2 cells are: edit and popovermenu. In files view the view is more or less the same and it is still a table.

If current view should be a grid -> i will create a grid ;) But i want to make sure that similar views are consistent to each other.

**Comment by michaelnissenbaum** [comment]
@JuliaKirschenheuter  I would say we have a situation where the solution lies somewhere in the middle. But I would say we can also implement it as a normal table instead of a grid.

**Comment by JuliaKirschenheuter** [comment]
https://github.com/nextcloud/server/pull/38424

**Comment by Pytal** [comment]
Please verify the table semantics on https://try.nextcloud.com/ltd/a11y/index.php/settings/users @michaelnissenbaum and let us know if adjustments should be made :)

**Comment by michaelnissenbaum** [comment]
Hi @JuliaKirschenheuter. You're on the right track, BUT in the case of tables, it is important not to mix ARIA roles and HTML elements, which you have done here. Either the table should be implemented entirely with ARIA roles or solely with HTML elements, without mixing them. You can find a good description at https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Table_Role.
![CleanShot 2023-07-19 at 12 48 37@2x](https://github.com/nextcloud/server/assets/48948959/b03b1fa3-da9b-4c1e-9aab-dd8e87620f63)

**Comment by Pytal** [comment]
The usage of both HTML elements and roles in the markup here is a technical limitation and did not seem optimal at first but given that browser engines interpret these semantics correctly i.e.

![image](https://github.com/nextcloud/server/assets/24800714/9c549975-ae95-487c-9c92-eddb217379f3)
> (from Firefox devtools)

and as a result provides the correct information to screenreaders, is this an acceptable solution @michaelnissenbaum?

**Comment by michaelnissenbaum** [comment]
Hi @JuliaKirschenheuter. Unfortunately, with such a mix of HTML elements and ARIA roles, it cannot be guaranteed that all screen readers will correctly render the table across all browsers. I would strongly advise against using such a mixed approach in this case.

**Comment by JuliaKirschenheuter** [comment]
Thanks a lot @michaelnissenbaum for this advice!

**Comment by Pytal** [comment]
Ultimately it's best to use HTML elements instead of hacking `vue-virtual-scroller` so I'll do it the in-house way à la https://github.com/nextcloud/server/blob/feat/f2v/files/apps/files/src/components/VirtualList.vue

**Comment by Pytal** [comment]
Any things we should take into account with VirtualList 👆 @skjnldsv? Shall we copy-n-adapt for users or abstract it into a reusable table component?

**Comment by skjnldsv** [comment]
Well, I don't know about the latest a11y hacks, but we're fully using official semantic for a table.
I don't have much time to dive into why we need rowgroups or such, there is probably some more changes needed (like scope attributes), but we should indeed take the latest VirtualList.vue component in https://github.com/nextcloud/server/pull/39808 yes :)

https://a11y-101.com/development/tables

**Comment by AndyScherzinger** [comment]
> Thanks a lot @michaelnissenbaum for this advice!
@michaelnissenbaum unfortunately we can't create a whole HTML table in that case because of technical limitations ;( Would it be ok if we would use only ARIA roles without HTML elements?

@michaelnissenbaum any comment on this and subsequent comments?

**Comment by michaelnissenbaum** [comment]
From my perspective, there's nothing preventing us from implementing the table entirely with ARIA. You can find the description here: https://www.w3.org/WAI/ARIA/apg/patterns/table/examples/table/.

**Comment by Pytal** [comment]
Let's see if the maintainer takes a look soon
- https://github.com/Akryum/vue-virtual-scroller/pull/825

If not, plan ARIA it is

**Comment by skjnldsv** [comment]
@Pytal you can also use the component Files is using.
https://github.com/nextcloud/server/blob/master/apps/files/src/components/VirtualList.vue

**Comment by AndyScherzinger** [comment]
@Pytal can this item be considered done for the moment or are there any open items? If So can you please add them to the top issue description? Thanks!
